specs/core/1_shard-data-chains.md

```diff
@@ -209,3 +272,275 @@ The `shard_chain_commitment` is only valid if it equals `compute_commitment(head
+    state.validator_balances[index] -= whistleblower_reward
+    validator.penalized_epoch = get_current_epoch(state)
+    validator.withdrawable_epoch = get_current_epoch(state) + LATEST_PENALIZED_EXIT_LENGTH
+```
+
+## Per-slot processing
+
+### Operations
+
+Add the following operations to the per-slot processing, in order the given below and _after_ all other operations (specifically, right after exits).
+
+#### Branch challenges
+
+Verify that `len(block.body.branch_challenges) <= MAX_BRANCH_CHALLENGES`.
+
+For each `challenge` in `block.body.branch_challenges`:
+
+* Verify that `slot_to_epoch(challenge.attestation.data.slot) >= get_current_epoch(state) - MAX_BRANCH_CHALLENGE_DELAY`.
+* Verify that `state.validator_registry[responder_index].exit_epoch >= get_current_epoch(state) - MAX_BRANCH_CHALLENGE_DELAY`.
+* Verify that `verify_slashable_attestation(state, challenge.attestation)` returns `True`.
+* Verify that `challenge.responder_index` is in `challenge.attestation.validator_indices`.
+* Let `depth = log2(next_power_of_two(SHARD_BLOCK_SIZE // 32 * EPOCH_LENGTH * (slot_to_epoch(challenge.attestation.data.slot) - challenge.attestation.latest_crosslink.epoch)`. Verify that `challenge.data_index < 2**depth`.
+* Verify that there does not exist a `BranchChallengeRecord` in `state.validator_registry[challenge.responder_index].open_branch_challenges` with `root == challenge.attestation.data.shard_chain_commitment` and `data_index == data_index`.
+* Append to `state.validator_registry[challenge.responder_index].open_branch_challenges` the object `BranchChallengeRecord(challenger_index=get_beacon_proposer_index(state, state.slot), root=challenge.attestation.data.shard_chain_commitment, depth=depth, inclusion_epoch=get_current_epoch(state), data_index=data_index)`.
+
+**Invariant**: the `open_branch_challenges` array will always stay sorted in order of `inclusion_epoch`.
+
+#### Branch responses
+
+Verify that `len(block.body.branch_responses) <= MAX_BRANCH_RESPONSES`.
+
+For each `response` in `block.body.branch_responses`:
+
+* Find the `BranchChallengeRecord` in `state.validator_registry[response.responder_index].open_branch_challenges` whose (`root`, `data_index`) match the (`root`, `data_index`) of the `response`. Verify that one such record exists (it is not possible for there to be more than one), call it `record`.
+* Verify that `verify_merkle_branch(leaf=response.data, branch=response.branch, depth=record.depth, index=record.data_index, root=record.root)` is True.
+* Verify that `get_current_epoch(state) >= record.inclusion_epoch + ENTRY_EXIT_DELAY`.
+* Remove the `record` from `state.validator_registry[response.responder_index].open_branch_challenges`
+* Determine the proposer `proposer_index = get_beacon_proposer_index(state, state.slot)` and set `state.validator_balances[proposer_index] += base_reward(state, index) // MINOR_REWARD_QUOTIENT`.
+
```

Comment:
Should we penalize the `challenger` a bit here? 
To reduce incentive to grief validators from being able to withdraw by keeping the `len(open_branch_challenges) > 0`

Reply:
I would say no. Otherwise we run into fisherman's dilemma issues. I think the goal should be that even if challengers find nothing worth challenging, they just query for random branches; this keeps this part of the system being tested and keep validators on their toes.Also there is a limit to how long each validator can be griefed because of `MAX_BRANCH_CHALLENGE_DELAY`.